File: Dart/14.longest_common_prefix.dart
void main() {
  print(longestCommonPrefix(["flower", "flow", "flight"]));
  print(longestCommonPrefix(["dog", "racecar", "car"]));
}

String longestCommonPrefix(List<String> strs) {
  var str = '';
  var index = 0;

  while (true) {
    final c = strs
        .map((e) => e.length > index ? e[index] : null)
        .reduce((a, b) => a == b && a != null ? a : null);

    if (c != null) {
      str = str + c;
      index++;
    } else {
      break;
    }
  }

  return str;
}

File: src/14.longest_common_prefix.kt
fun main() {
    println(longestCommonPrefix(arrayOf("flower", "flow", "flight")))
    println(longestCommonPrefix(arrayOf("dog", "racecar", "car")))
}

fun longestCommonPrefix(strs: Array<String>): String {
    var str = ""
    var index = 0

    while (true) {
        val c: String = strs.map { e -> if (e.length > index) "${e[index]}" else "" }
            .reduce { a, b -> if (a == b && a != "") a else "" }

        if (c.isNotEmpty()) {
            str += c
            index++
        } else {
            break
        }
    }

    return str
}